src/NDItk/multigroup/Metadata.hpp

```diff
@@ -113,16 +148,43 @@ class Metadata {
   template< typename Iterator >
   void read( const std::string& keyword, Iterator& iter, const Iterator& end ) {
 
-    if      ( keyword == this->zaid_.keyword() )           { readRecord( this->zaid_, iter, end ); }
-    else if ( keyword == this->library_name_.keyword() )   { readRecord( this->library_name_, iter, end ); }
-    else if ( keyword == this->source_date_.keyword() )    { readRecord( this->source_date_, iter, end ); }
-    else if ( keyword == this->process_date_.keyword() )   { readRecord( this->process_date_, iter, end ); }
-    else if ( keyword == this->awr_.keyword() )            { readRecord( this->awr_, iter, end ); }
-    else if ( keyword == this->atomic_weight_.keyword() )  { readRecord( this->atomic_weight_, iter, end ); }
-    else if ( keyword == this->temperature_.keyword() )    { readRecord( this->temperature_, iter, end ); }
-    else if ( keyword == this->dilution_.keyword() )       { readRecord( this->dilution_, iter, end ); }
-    else if ( keyword == this->groups_.keyword() )         { readRecord( this->groups_, iter, end ); }
-    else if ( keyword == this->reactions_.keyword() )      { readRecord( this->reactions_, iter, end ); }
+    if      ( keyword == this->zaid_.keyword() )          { readRecord( this->zaid_, iter, end ); }
+    else if ( keyword == this->library_name_.keyword() )  { readRecord( this->library_name_, iter, end ); }
+    else if ( keyword == this->source_date_.keyword() )   { readRecord( this->source_date_, iter, end ); }
+    else if ( keyword == this->process_date_.keyword() )  { readRecord( this->process_date_, iter, end ); }
+    else if ( keyword == this->awr_.keyword() )           { readRecord( this->awr_, iter, end ); }
+    else if ( keyword == this->atomic_weight_.keyword() ) { readRecord( this->atomic_weight_, iter, end ); }
+    else if ( keyword == this->temperature_.keyword() )   { readRecord( this->temperature_, iter, end ); }
+    else if ( keyword == this->dilution_.keyword() )      { readRecord( this->dilution_, iter, end ); }
+    else if ( keyword == this->reactions_.keyword() )     { readRecord( this->reactions_, iter, end ); }
+    else if ( keyword.find( this->primary_groups_.keyword() ) == 0 ) {
+
+      if ( keyword == this->primary_groups_.keyword() ) {
+
+        readRecord( this->primary_groups_, iter, end );
+      }
+      else {
+
+        base::Keyword secondary( keyword );
+        auto pos = std::lower_bound( this->outgoing_groups_.begin(),
```

Comment:
As before, what exactly is this doing? I see the readRecord later with *pos. Is this all to find where the actual energy group boundaries start?

Reply:
As mentioned above, lower_bound algorithm returns the position of the element that is either the one we are looking for or the first element that is larger than the one we are looking for (if that element is not there). It's basically binary search. This is why I actually check if the element at pos is the one with the particle type we are looking for - which would mean that we have a duplicate keyword. Once we know there is no duplicate, we insert a new record in that position.

This basically keeps the vector sorted by particle number while inserting records. This is more efficient compared to just using push_back and then applying sort.